### PR TITLE
CB-17388 Fix infinite loop in case of custom image not found and disable change catalog for custom image catalog

### PR DESCRIPTION
**Pull request by pkedvessy**
See detailed description in the commit message.

### CODE REVIEW

**Comment by akanto** [comment]
This should be 2.57.0 since it is broken on prod as we speak.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/StackImageService.java [hunk 1]
```diff
@@ -93,7 +93,18 @@ public Image getCurrentImage(Stack stack) throws CloudbreakImageNotFoundExceptio
...
+            ImageCatalog sourceImageCatalog = imageCatalogService.getImageCatalogByName(stack.getWorkspace().getId(), currentImage.getImageCatalogName());
+            if (imageCatalogService.isCustomImageCatalog(sourceImageCatalog)) {
+                throw new CloudbreakServiceException(
+                        String.format("Changing custom image catalog '%s' is not supported.", sourceImageCatalog.getName()));
+            }
+
             ImageCatalog targetImageCatalog = imageCatalogService.getImageCatalogByName(stack.getWorkspace().getId(), imageCatalog);
+            if (imageCatalogService.isCustomImageCatalog(targetImageCatalog)) {
+                throw new CloudbreakServiceException(
+                        String.format("Custom image catalogs are not supported to replace the current '%s' image catalog.", sourceImageCatalog.getName()));
```
Comment (TheTinkerDad) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/stack/StackImageService.java:105]: This error message is misleading - a custom JSON image catalog is still a custom image catalog. There should be something like "only JSON catalogs are supported" or something like that.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/stack/StackImageService.java [hunk 2]
```diff
@@ -93,7 +93,18 @@ public Image getCurrentImage(Stack stack) throws CloudbreakImageNotFoundExceptio
     public void changeImageCatalog(Stack stack, String imageCatalog) {
         try {
             Image currentImage = componentConfigProviderService.getImage(stack.getId());
+            ImageCatalog sourceImageCatalog = imageCatalogService.getImageCatalogByName(stack.getWorkspace().getId(), currentImage.getImageCatalogName());
+            if (imageCatalogService.isCustomImageCatalog(sourceImageCatalog)) {
+                throw new CloudbreakServiceException(
+                        String.format("Changing custom image catalog '%s' is not supported.", sourceImageCatalog.getName()));
```
Comment (TheTinkerDad) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/stack/StackImageService.java:99]: Misleading error message - a custom catalog is a custom catalog, it's format is irrelevant. This should be something like "migrating from a non-JSON based catalog to a JSON based one is not supported" or something like that.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/image/ImageCatalogService.java [hunk 3]
```diff
@@ -806,11 +806,13 @@ private StatedImage getCustomStatedImage(ImageCatalog imageCatalog, String image
             StatedImage sourceImage = getSourceImageByImageType(customImage);
             LOGGER.info("Custom image '{}' is a {} image '{}' customization.", imageId, customImage.getImageType(), customImage.getCustomizedImageId());
             return customImageProvider.mergeSourceImageAndCustomImageProperties(
-                        sourceImage, customImage, imageCatalog.getImageCatalogUrl(), imageCatalog.getName());
+                    sourceImage, customImage, imageCatalog.getImageCatalogUrl(), imageCatalog.getName());
+        } else if (Strings.isNullOrEmpty(imageCatalog.getImageCatalogUrl())) {
+            throw new CloudbreakImageNotFoundException(
+                    String.format("Could not find any custom image with id: '%s' in catalog: '%s'", imageId, imageCatalog.getName()));
```
Comment (TheTinkerDad) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/image/ImageCatalogService.java:812]: DB based custom image catalogs don't have a URL by nature. If the image is within such catalog, the error message is problematic - imageID will match, catalog name will match, yet the response will still say "could not find any image", which is confusing.

**Review by TheTinkerDad — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/sequenceiq/cloudbreak/service/image/ImageCatalogService.java (hunk 3 above)
Comment (pkedvessy) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/image/ImageCatalogService.java:812]: Already discussed on slack:
It is not related to the bug but unit tests highlighted that null imageCatalogUrl will cause an NPE. This is a generic public method to lookup an image so having this error message makes sense.

Comment (pkedvessy) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/image/ImageCatalogService.java:812]: In terms of changing the image catalog, this method is used to validate that the target image catalog includes the image (by uuid) what we already use from the source catalog.

**Review by TheTinkerDad — APPROVED** [review]

**Review by pdarvasi — APPROVED** [review]

**Pull request merged**